### ISSUE TITLE
Noticeably lower quality than x264 at same bitrate, can't push it higher

### ISSUE THREAD

**Issue by c2h5oh** [Xaymar/OBS-AMD-Advanced-Media-Framework#116]
I wouldn't be surprised if that's AMF / VCE limitation, but I'm reporting it in case it isn't.

tl;dr; Regardless of settings selected recording quality at 3500 CBR for 720p video roughly matches x264 superfast preset, which is considered too low quality for vast majority of uses.
veryfast preset at those settings costs just 5% extra cpu load.

OBS: 0.16.2 64bit
Plugin: 1.3.3.1
Fury + i7 6700k

I've recorded the very same source: Bioshock Infinite benchmark running in 1080p with ultra settings, with forced vsync - this was scaled down to 720p and recorded at 30 and 60 fps. 
All AMF recordings used Twitch preset, with bitrate increased to 3500kbps (or other rate control method which resulted in average bitrate being around 3500kbps) + quality / profile changed - see video titles

Videos can be viewed here: https://vimeo.com/user54378907/videos

Source files can be found here: https://drive.google.com/drive/folders/0B2dZ43S2E-izQ09SQmFSTEZXaU0?usp=sharing (also includes high bitrate file for comparison)

**Comment by DoctorSoup143** [comment]
Encountering the same issue. I've tried tons of settings and still haven't found a fix.

**Comment by kytos22** [comment]
Well, you usually need +500kb to make a hardware encoder look like pure cpu x264.https://github.com/Xaymar/OBS-AMD-Advanced-Media-Framework/wiki/Frequently-Asked-Questions#q-how-do-i-make-it-look-like-x264

Also make sure you have the latest drivers.

This will change when youtube, twitch and such start using x265 where you will need much more cpu power. (better to use hardware encoding than waste cpu cycles)

**Comment by Benman2785** [comment]
its normal - but we can try to improve it - Maintainer Edit: Please do not advise users to do bad things. Thanks! ^Xaymar

**Comment by Xaymar** [comment]
Odd, my email responses don't show up. Anyway, try increasing B-Pictures to 3, enable B-Picture Reference,  Disable HRD Compatibility and Disable Frame Skipping. If you're on 1.3.3.1, you might also need to set Minimum QP to 0 and Maximum QP to 51 in VBR_LAT - i left bit of test code in.

**Comment by Benman2785** [comment]
i will try ;)

btw its not bad - you only have to know how h264 works :p

**Comment by Benman2785** [comment]
Xaymar - ur code in VBR_Lat fixed the highpeak on that scene - but i got quiet a lot of encoder overloads

![stream-1477306761](https://cloud.githubusercontent.com/assets/21282980/19643589/5a87547a-99eb-11e6-9ae2-a287da40595f.jpg)
-> u see no peak after that black-scene with no movement // still waiting for AMDs CBR fix

**Comment by Xaymar** [comment]
VBR_LAT is not a replacement for CBR

**Comment by Benman2785** [comment]
i know - but it doesnt have that bug (high peak) while still haveing the bug (low bitrate)

**Comment by Xaymar** [comment]
That's because  it isn't CBR

**Comment by c2h5oh** [comment]
Minimum QP seems to be bugged (or at least was the last I tried) - if set to 0 quality of high motion / dark scenes was horrible (with bitrate dropping to 10~15% of target) - super odd, with min QP being the upper bound. Setting it to 1 seemed to fix it (are you comparing bool to int somewhere?).

Anyway I'll re-record the same scene later today with different settings and let's do some comparisons - let's build HQ versions of the profiles

**Comment by Benman2785** [comment]
@c2h5oh that is caused by AMDs new AMF -> will be fixed soon

**Comment by Xaymar** [comment]
Benman, your opinion is nice and all but please stop sharing it.

The advanced options can't "improve" quality over what AMF has to offer and
you've so far been ignoring most information i try to get in your head.
It's frustrating and anything but helpful.

There is a reason i don't support people who use the experimental
parameters of the [Advanced] interface.

Always use presets unless you absolutely know what you are doing, have no
doubts, did your research and not just followed some dumb online tutorial,
video or reddit post. They exist for the reason to make things easy to use
and set up things according to the hardware available.

**Comment by Xaymar** [comment]
Holy shit that appeared late

**Comment by Xaymar** [comment]
Set B-pictures to 3, it might help a little.

**Comment by Xaymar** [comment]
I did some tests and the results are not promising at all. Often AMF compresses the initial NV12 image so much that color information is actually lost and it ends up grey. Then there is the issue of the I-Frame QP being too high for that bitrate. X264 picked an appropriate overall bitrate of 3320 kbit without filler data while AMF went as low as 1500 - on a static image that contains a lot of bright/dark mixture. AMF looks significantly worse than X264 at the fastest setting.

I request that @MikhailAMD comment here and explain how to get better quality out of AMF without having to do additional reverse engineering. Do we need full range encoding, 2-pass encoding, ...?

**Comment by MikhailAMD** [comment]
1. Please provide GPU model, driver version and OS so we could reproduce. 
2. New version of AMF (1.3) is identical from encoder perspective to Media SDK 1.1. The change is mostly in object factory and bug fixes. 
3. There are two problems reported by @Xaymar and related to CBR: 
   - slow encoding on CBR
   - bad quality on VCE2. 
     They both should be fixed in the upcoming driver release.
4. HW encoder cannot match quality of software encoder. The goal is speed and CPU offload. We have had several projects with other software vendors and it does make sense to compare HW with HW only. 
5. Pre-pass would help, working on it, but I cannot comment on the date of publishing. Note, that pre-pass is not a full 2-pass since scanning and bitrate redistribution is limited to one frame. 
6. Once I have the fixes (they are not in AMF that I own but in FW and the driver) I will do some experiments and check if quality can be improved.
7. I don't think reverse engineering would give some improvements, we are not trying to hide an important parameter though feel free to try.

**Comment by kytos22** [comment]
@MikhailAMD I hope we can get HEVC/VP9 soon so we can use it to test and local recording for now.

**Comment by Xaymar** [comment]
1. ASUS R9 285 Strix II OC (AMD Radeon R9 285), 16.10.2, Windows 10 64-bit
2. There seems to be a regression in the driver then, since Media SDK was mostly bug free - except for SubmitInput long submission time stuff.
3. I will be reporting another one, but I'm not sure if it can be fixed: QP values are picked too high for I-Frames, reducing overall quality a lot - see previous example.
4. As far as my understanding of Hardware and Software goes, Hardware Encoders can match the quality of Software encoders if they are designed to. I do not know how the internals of the VCE core work (yet) so I can't comment on any changes that might need to be done or similar.
5. Alright
6. Okay, we'll be waiting.
7. Reverse Engineering gave me control over CABAC, Aspect Ratio and GOP Size, which some people seem to actually need for better quality. So if there are any more parameters that I can access without writing my own VCE driver interface, they could prove useful for quality.

**Comment by Benman2785** [comment]
@Xaymar @MikhailAMD 

thank you both for your dedication in trying to fix all stuff ;)

to 4. as far as i know HW and SW does different magic and cant match with same bitrate - but usually 200-300kbps are needed to look similar with 720p60 (what most would appreciate)

to 7. GOP-Size is highly important to ensure that especially youtube wont take long to process a video (as they wish a GOP of 1/2 FPS) -> saves time when u upload/release a 2h video // AspectRatio is a need for videos ;)

but after all: this Xaymars VCE Plugin for OBS Studio looks pretty good now -> better then old Jackuns OBS (classic) fork -> so there was improvements made by both of u

sry for bitting someons nerves

**Comment by MikhailAMD** [comment]
- We always welcome constructive comments :)
- General parameter statement: Please tell me why do you need a parameter (a use case) and I will work with codec team to expose it if it is possible. It is true for hidden or non-existent ones.
- Regarding GOP size parameter: It was redundant since the same thing can be controlled by  IDR_PERIOD. Having two of them created complicated logic of I-frame appearance. Now it is retired and is not used in the driver.  
- CABAC does not affect quality of the picture - it is just a different lossless method of packing data bits after quantization (entropy encoding). Though, it may have better compression ration compare to CALV. I will expose it in the next AMF release. 
- Aspect ratio: it is VUI optional parameter and does not affect encoding. Codec team is OK to expose it.

**Comment by Benman2785** [comment]
ok - i wont share my opinion

but usually i know what im doing ;)

und wir können auch deutsch schreiben
falls ich mal für dich was testen soll sag bescheid

Am 25-Oct-16 um 20:17 schrieb Michael Fabian Dirks:
Benman, your opinion is nice and all but please stop sharing it.

The advanced options can't "improve" quality over what AMF has to offer and
you've so far been ignoring most information i try to get in your head.
It's frustrating and anything but helpful.

There is a reason i don't support people who use the experimental
parameters of the [Advanced] interface.

Always use presets unless you absolutely know what you are doing, have no
doubts, did your research and not just followed some dumb online tutorial,
video or reddit post. They exist for the reason to make things easy to use
and set up things according to the hardware available.

**Comment by c2h5oh** [comment]
@MikhailAMD

> CABAC does not affect quality of the picture - it is just a different lossless method of packing data bits after quantization (entropy encoding). Though, it may have better compression ration compare to CALV. I will expose it in the next AMF release.

Better, lossless compression -> lower bitrate at exact same quality OR more bandwidth available to lossy component of compression while maintaining the bitrate. 
Given that bitrate is the usual constraint when streaming effectively it does in fact affect quality of the picture.

> HW encoder cannot match quality of software encoder. The goal is speed and CPU offload. We have had several projects with other software vendors and it does make sense to compare HW with HW only.

VCE3 is supposed to be able to handle 60fps RGB at 2160p. At 720p (89% fewer pixels per frame), NV12 (half of the bytes per pixel), 30 fps (half of the frame rate) it has to handle just 2.8% that bandwidth. I know that VCE being a dedicated unit is not as flexible as software running on a general purpose cpu, but my assumption is that there is some room to improve quality at lower resolutions.

Feature request time:
- Constant Rate Factor

**Comment by Xaymar** [comment]
@MikhailAMD 
My bad with GOP Size then, it still worked here but that could've been before I installed 16.10.2. I'll remove it in a future plugin release. Could I manage to make a GOP Size of half the framerate through other means then? Header Insertion Spacing perhaps?

_CABAC_ is actually recommended for most streams, since you can usually pack ~5% more bits into the stream on average.  So the extra 200-300 kbit/s per second would actually end up being nearly the same exact bitrate as you would need for near-identical software encoding quality.

I don't know about other parameters that need to be exposed yet. If there are some, I will open up an issue for it on the AMF issue tracker.

@Benman2785 Ich habe deine Antwort auf die extrem verzögerte E-Mail Antwort von mir entfernt. Tut mir Leid wegen der miesen Reaktion, ich war an dem Tag sehr genervt davon dass du wiederholt experimentelle Parameter genutzt hast ohne aktuell zu wissen ob diese wirklich funktionieren - und dann auch noch andere dazu aufrufen wolltest diese zu verwenden.

**Comment by MikhailAMD** [comment]
You can put any value to IDR_PERIOD to control GOP size. GOP is the size between I or IDR frames. Half of the frame rate is fine.  HEADER_INSERTION_SPACING controls SPS/PPS insertion, not GOP size. Normally these two parameters should be set the same. HEADER_INSERTION_SPACING makes sense to use if you are using intra-refresh and there are no I or IDR frames.

**Comment by Shivansps** [comment]
What about more of the high end of the bitrate scale? Personally im having problems trying to get anywhere near the quality/bitrate ratio that i used to get with my old NV card using NVENC, now i have RX480 (VCE 3.0), and the results so far are more than disappointing... Using OBS with Xaymar plugin im getting the best results, so far, but they are still VERY, VERY far, i used to record 1080p@35mbps with shadowplay, i need about 70mbps to match the results, and that is kinda insane.

And, not related to this plugin, but maybe usefull to MikhailAMD, other programs, like Plays.TV that used to be bundled with drivers, has VBR completely broken, like i set a manual bitrate of 35mbps, but in the result video the overall bitrate may end up as low as 10mbps with absolutely terrible results. Ive been talking about all this in here: https://forums.anandtech.com/threads/awful-amd-vce-performance.2489847/ and i provided sample videos.

RX480 8GB
I5-2500K
16.10.2

**Comment by Xaymar** [comment]
I managed to get almost better quality out of AMF than x264 veryfast. AMF still has trouble with really dark colors (cut off by Partial NV12 color space) and cross-block movement, but otherwise this looks much better and takes much less CPU than x264. Perhaps @MikhailAMD can tell me if better movement prediction and full range encoding are possible in hardware and could come in a future driver upgrade.

Here is the configuration, the trick seems to be to set Min/Max QP to be as loose as possible and keep VBV Buffer really strict:
- Requires plugin version 1.4.0.0 (to be released publicly).
- ---
- Usage: Transcoding
- Profile: High
- Profile Level: 5.2
- Maximum LTR Frames: 0
- ---
- Rate Control Method: CBR
- Bitrate Target/Peak: 3500/3500
- Min/Max QP: 0/51
- I-/P-/B-Frame QP: 0/0/0
- Ref/Non-Ref B-Picture Delta QP: -10/-10
- VBV Buffer: Automatic
- VBV Buffer Strictness: 100%
- VBV Buffer Fullness: 100%
- Maximum Access Unit Size: 0
- Filler Data: Disabled
- Frame Skipping: Disabled
- Enforce HRD Compatibility: Disabled
- Keyframe Interval: 1
- Header Insertion Spacing: 0
- B-Picture Pattern: 3
- B-Picture Reference: Enabled
- Deblocking Filter: Enabled
- Slices Per Frame: 0
- Intra-Refresh Num MBs Per Slot: 0
- Quality Preset: Quality
- Scan Type: Progressive
- Motion Estimation: Both
- CABAC: Enabled

Examples files: [AMF](http://cdn.xaymar.com/private/2016/10/2016-10-27%2004-52-01_amf.mkv), [X264](http://cdn.xaymar.com/private/2016/10/2016-10-27%2005-21-44_x264.mkv).

**Comment by Benman2785** [comment]
@MikhailAMD 
shouldnt GOP-Size be number of frames grouped -> everything including i- p- b- frames? 
ps would be nice to manually set fixed i-frames in frames (but also have an auto function) -> so we can play around (improve) with compression for different scenes/games regarding their screen-movement

@Shivansps 
mh, i think you do something wrong. with Xaymars VCE Plugin i get pretty good quality with even low bitrate (3500kbps for stream) in 720p30 // 1080p60 should look decent with 35mbps even with lots of movement -> maybe you can join my teamspeak server and i will help you set it correctly

@Xaymar 
schon ok - kann dich verstehen // dennoch mach weiter so ;)

**Comment by MikhailAMD** [comment]
Here is the definition of GOP: https://en.wikipedia.org/wiki/Group_of_pictures
There are group of the encoder properties at the end of the header file that have to be set on input surface. Using them you can force the submitted surface to be of any type: I, P or B, independent from automatic.

**Comment by chummy02** [comment]
Benman GOP size is defined by one I-frame only at beginning and the other frames like you said form the GOP. Everytime one I-frame is added there a new GOP. So when you define IDR Period(keyframe interval) of 60 for a 60fps video you displaying one I-frame every 60 frames what already is forming your GOP size. So there is nothing  happening when adding GOP size together to IDR Period since IDR writes the i-frames and decide the GOP size.
I-frames are already fixed when you set the keyframe interval, i think adaptive I-frames are not supported by VCE. If you want to try lower GOP size like Youtube half framerate GOP just set IDR Period(Keyframe interval) at half your recording framerate so this would do what you want to have a GOP of 30 for a 60fps video, or set IDR Period to 15 if your footage is 30fps. But too low GOP for local recording make bitrate skyrocket, i dont know how it work for streaming.

GOP size setting in Jackun fork and Xaymar is being always overwritten by Keyframe interval which is the one which defines the GOP size as already said. Take a look at this log from my recorded video with Xaymar plugin. Recorded with 60fps and IDR Period of 30, created a GOP size of 30.
[test.txt](https://github.com/Xaymar/OBS-AMD-Advanced-Media-Framework/files/557590/test.txt)

**Comment by wazerstar** [comment]
@Xaymar That is a good example video with AMF in dark areas,

**Comment by Benman2785** [comment]
@MikhailAMD @Xaymar 

is that implemented yet? so i would like to test a game with fixed i-Frames every 5 Pictures -> should result in not so good compression but good looking movement

@chummy02 
yeah - Mikhail said it is supported -> lets find out if we have improvements ;) // btw many iframes usually make it look better when heavy movement is on screen - but takes ~3x as much bitrate as p frames

**Comment by Xaymar** [comment]
It's always been implemented. An IDR-Period of 5 in the Advanced one would do exactly that. The simpler one uses Keyframe Interval (Seconds).

**Comment by Benman2785** [comment]
i will try some stuff with it ;)

**Comment by c2h5oh** [comment]
I'm trying to make my tests more scientific - the recordings I linked in the ticket have each been done manually, they are off by a couple of frames, varying length and etc.
1. I'm going to record a couple video clips with lossless compression - I'm selecting a source right now: I want scenes with varying levels of motion, light, and complexity, each lasting at least 5-10s, all within a single, short video (around 1 min would be best..). 720p and 1080p at 30 and 60 fps. 1440p if it's possible to record 1440p without a 1440p screen by using AMD super resolution.
2. I'll use mpv with hardware decoding disabled and OpenGL or DirectX output to simulate game-like source
3. I'll capture the video with OBS with different settings that we can use for comparisons.

I own: 
- 7870 (VCE 1.0)
- 390x (VCE 2.0)
- Fury (VCE 3.0)

So we can do a comparison between 3 out of 4 supported right now. I might be able to borrow a R9 470 for full set, but I'm not sure right now - if any of you guys live in Toronto and would like to lend me one for one evening that would be great. Other hardware is beefy enough to take CPU and disk bandwidth out of the equation.

I'm missing 2 things:
1. Settings combinations that I should be testing (setting1: value1, value2; setting2: value3, value4) etc  @Xaymar this is something you're best suited to provide
2. A way to sync start/stop of video playback with recording start/stop with a single frame precision if possible. Is there something that would be more reliable than faking hotkeys? mpv supports lua scripting. ideas?

I need to be able to automate the whole thing. 2 (or 3) resolutions \* 2 frame rates \* 4-6 bitrates is already 16 to 36 runs per setting set, per card. Even with 1 minute video it's going to be hours - I need to be able to start it and go to work.

I'll automate the comparison frames extraction too when it's done - we don't want to be watching the same 1 minute clip for 10 hours, do we? ;-)

**Comment by Xaymar** [comment]
@c2h5oh Best test cases across many cards?

Common Resolutions:
- 640x480
- 800x600
- 1024x768
- 1280x720
- 1366x768
- 1600x900
- 1920x1080

Common Framerates:
- 24
- 30
- 48
- 60

Other Settings:
- Preset: Indistinguishable or High Quality (Lossless is too hard for most systems, especially VCE1 cards)
- Profile: High
- Profile Level: Automatic
- Minimum QP: 1 (some cards have issues with 0)
- Maximum QP: 51
- VBV Buffer: Automatic (1.4.0.0 and up)
- VBV Buffer Strictness: 100%
- VBV Buffer Fullness: 0%
- Frame Skipping: Disabled
- Enforce HRD Compatibility: Disabled
- Keyframe Interval: 2.0
- Deblocking Filter: Enabled
- Motion Estimation: Half- & Quarter-Pixel
- CABAC: Disabled
- Memory Type: Host or DirectX 9/11, whatever works best
- Compute Type: None (Host) or OpenCL (DirectX)

As for 2, I have no idea. AutoHotkey is your best bet here most likely.

**Comment by Xaymar** [comment]
Quoting irc:

> [23:57:49] <Sapiens|Work> !params
> [23:57:49] <OBScommits> For OBS Studio, --collection [scene collection name], --profile [profile name], --scene [scene name], --startstreaming, --startrecording. If a collection/profile/scene has a space in the name, use quotes around the full name. For portable mode, read http://github.com/jp9000/obs-studio/releases/tag/0.11.1 | For OBS Classic, read http://jp9000.github.io/OBS/general/shortcut.html

**Comment by chummy02** [comment]
Since VCE 1 and 2 dont support b-frames would be interesting to test VCE3+ without b-frames to compare if there were some quality improvement beside use of b-frames. So make test with

My VCE 2 is worse than x264 Ultrafast while my Quicksync 3.0 is draw to x264 veryfast. NVENC 1.0 is close to Quicksync 3 too and both NVENC1 and QS3 were launched before VCE2. Apparently AMD has the worse ASIC quality/speed of three. Just wanted to see if VCE 3 at least match Quicksync quality.

**Comment by Xaymar** [comment]
VCE 2 has a bug with B-Pictures which is supposedly being fixed. Some day. Maybe next year when HEVC is out.

**Comment by chummy02** [comment]
VCE2 has almost 3 years launch and b-frames are broken since then. VCE2 cannot do Quality encoding at 1080p@60 too, for quality the max FPS is around 45 at 1080p. VCE2 is worse at IQ and performance compared to older Quicksync and NVENC. A GTX650 has better encoder quality and performance than a 290/390. AMD dont put much effort in their ASIC's encoders like Intel and NVIDIA do from hardware capabilities to software development.

**Comment by Xaymar** [comment]
I removed the Off-Topic stuff. If you want to discuss Nvidia, Intel and AMD differences and how bad each one is use a forum for that and not the issue tracker.

**Comment by Xaymar** [comment]
There has been no activity here for 4 days and since this is not technically an issue with the plugin I will close this.

**Comment by MikhailAMD** [comment]
@c2h5oh I can help you with RX480 or 470 card. Will you be able to visit AMD office in Markham in a few days? I am waiting for shipment with cards from the US.

**Comment by kytos22** [comment]
Sorry for my interruption but @MikhailAMD you can maybe help @Xaymar with a RX470/80 in Germany so he can test and improve the plugin with the new Polaris capabilities. :)

**Comment by MikhailAMD** [comment]
I thought @Xaymar has it and @c2h5oh asked for help in Toronto area where I am. If not, certainly I will try to help. I have approval for 1 board for now. Pls let me know.

**Comment by kytos22** [comment]
Well, he had one but it broke and he returned it. If you can, i think is a good opportunity not only for now but for future updates. 
Am I right @Xaymar?

**Comment by Xaymar** [comment]
@MikhailAMD @M4RK22 
The RX 480 I had was pretty much Dead-On-Arrival, only after two days I managed to get it working and then with massive artifacting. I sent it in for replacement but the replacement card also had this issue, so I gave it back and spent the money on a better air+water cooling solution and a non-broken computer case instead.
I'd be grateful for any card that you could offer so I can test on a GCN4/VCE3.4 card as well, my address is as always listed at the bottom of [the About Me page on my Website](https://www.xaymar.com/about-me/). Edit: I can also travel some amount of distance in case you want me to show up at an AMD building.

**Comment by MikhailAMD** [comment]
I sent you a private email for the postal address.

**Comment by Xaymar** [comment]
Thank you, I'll be waiting on the E-mail so I can reply.

**Comment by MikhailAMD** [comment]
I sent email to info@xaymar.com more then an hour ago.

**Comment by Xaymar** [comment]
It seems to have been dropped by the spam filter for some reason then (can't find any log entry though), can you send it to my private address instead? michael.fabian.dirks@gmail.com

Edit: updated the e-mail listed on my profile for the time being until I can figure out what happened to the email you sent off.
